File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/index/IndexWriters.java
package ai.platon.pulsar.crawl.index;

import ai.platon.pulsar.common.Strings;
import ai.platon.pulsar.common.config.ImmutableConfig;
import org.slf4j.Logger;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Creates {@link IndexWriter} implementing plugins.
 */
public class IndexWriters implements AutoCloseable {

    Logger LOG = IndexWriter.LOG;

    private ImmutableConfig conf;
    private ArrayList<IndexWriter> indexWriters = new ArrayList<>();

    public IndexWriters() {
    }

    public IndexWriters(ImmutableConfig conf) {
        this(Collections.emptyList(), conf);
    }

    public IndexWriters(List<IndexWriter> indexWriters, ImmutableConfig conf) {
        this.conf = conf;
        this.indexWriters.addAll(indexWriters);
    }

    public void open() {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.open(conf);
            } catch (Throwable e) {
                LOG.error("Failed to open indexer. " + Strings.stringifyException(e));
            }
        }
    }

    public void open(String indexerUrl) {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.open(indexerUrl);
            } catch (Throwable e) {
                LOG.error("Failed to open indexer. " + Strings.stringifyException(e));
            }
        }
    }

    public void write(IndexDocument doc) {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.write(doc);
            } catch (Throwable e) {
                LOG.error("Failed to write indexer. " + Strings.stringifyException(e));
            }
        }
    }

    public void update(IndexDocument doc) {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.update(doc);
            } catch (Throwable e) {
                LOG.error("Failed to update indexer. " + Strings.stringifyException(e));
            }
        }
    }

    public void delete(String key) {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.delete(key);
            } catch (Throwable e) {
                LOG.error("Failed to delete indexer. " + Strings.stringifyException(e));
            }
        }
    }

    @Override
    public void close() {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                // log.info("[Destruction] Closing IndexWriter " + indexWriter.getName() + ", ...");
                indexWriter.close();
            } catch (Throwable e) {
                LOG.error("Failed to close IndexWriter " + indexWriter.getName());
                LOG.error(Strings.stringifyException(e));
            }
        }

        indexWriters.clear();
    }

    public void commit() {
        for (IndexWriter indexWriter : indexWriters) {
            try {
                indexWriter.commit();
            } catch (Throwable e) {
                LOG.error("Failed to commit indexer. " + Strings.stringifyException(e));
            }
        }
    }

    @Override
    public String toString() {
        return indexWriters.stream().map(n -> n.getClass().getName()).collect(Collectors.joining(", "));
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/index/IndexWriters.kt
package ai.platon.pulsar.crawl.index

import ai.platon.pulsar.common.config.ImmutableConfig
import ai.platon.pulsar.common.getLogger
import ai.platon.pulsar.common.stringify

/**
 * Creates [IndexWriter] implementing plugins.
 */
class IndexWriters(
    val indexWriters: MutableList<IndexWriter> = mutableListOf(),
    val conf: ImmutableConfig
) : AutoCloseable {
    var LOG = getLogger(IndexWriters::class)

    constructor(conf: ImmutableConfig) : this(mutableListOf(), conf)

    fun open() {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.open(conf)
            } catch (e: Throwable) {
                LOG.error("Failed to open indexer. " + e.stringify())
            }
        }
    }

    fun open(indexerUrl: String?) {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.open(indexerUrl)
            } catch (e: Throwable) {
                LOG.error("Failed to open indexer. " + e.stringify())
            }
        }
    }

    fun write(doc: IndexDocument?) {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.write(doc)
            } catch (e: Throwable) {
                LOG.error("Failed to write indexer. " + e.stringify())
            }
        }
    }

    fun update(doc: IndexDocument) {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.update(doc)
            } catch (e: Throwable) {
                LOG.error("Failed to update indexer. " + e.stringify())
            }
        }
    }

    fun delete(key: String) {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.delete(key)
            } catch (e: Throwable) {
                LOG.error("Failed to delete indexer. " + e.stringify())
            }
        }
    }

    override fun close() {
        for (indexWriter in indexWriters) {
            try {
                // log.info("[Destruction] Closing IndexWriter " + indexWriter.getName() + ", ...");
                indexWriter.close()
            } catch (e: Throwable) {
                LOG.error("Failed to close IndexWriter " + indexWriter.name)
                LOG.error(e.stringify())
            }
        }
        indexWriters.clear()
    }

    fun commit() {
        for (indexWriter in indexWriters) {
            try {
                indexWriter.commit()
            } catch (e: Throwable) {
                LOG.error("Failed to commit indexer. {}", e.stringify())
            }
        }
    }

    override fun toString(): String {
        return indexWriters.joinToString { it.javaClass.name }
    }
}
